### ISSUE TITLE
Invalid simplification

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#2694]
The IDE incorrectly considers cast (uint) to be unnecessary in this case.

```C#
class Program
{
    static void Main()
    {
        int a = 0;
        long l = 0;
        long x = (uint)a | l;
    }
}
```

Without the cast the compiler reports a warning:
```
warning CS0675: Bitwise-or operator used on a sign-extended operand; consider casting to a smaller unsigned type first
```

**Comment by jaredpar** [comment]
Duplicate of #4027.  Going to treat that as the primary.

**Comment by AlekseyTs** [comment]
I do not think this is a duplicate. The warning is also reported by VS2013 compiler in this scenario.

**Comment by Pilchie** [comment]
Looks like @DustinCampbell already fixed this in #4643

**Comment by Pilchie** [comment]
@DustinCampbell Can you verify?

**Comment by DustinCampbell** [comment]
Yes, this is fixed on recent bits with #4643.
